### PR TITLE
Feature/otp login

### PR DESCRIPTION
**Pull request by ibilous**
Add possibility login to the system over OTP. Current implementation makes OTP as oauth server that returns login page and provides necessary communication with UAA: creating jwt access token, verifying it, returning userInfo

Sequence diagram: https://i.imgur.com/W1Iae1e.png

Known issue: each time when user load login page it prints out html code to console in a RestLoggingAspect class.  I have not found how to prevent it. Should it be fixed in logging library or there is an easier solution ?

### CODE REVIEW

File: src/main/java/com/icthh/xm/ms/otp/repository/UaaRepository.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.stereotype.Component;
+import org.springframework.util.LinkedMultiValueMap;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.client.RestTemplate;
+
+import javax.annotation.PostConstruct;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.icthh.xm.ms.otp.config.Constants.ACCESS_TOKEN;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/otp/repository/UaaRepository.java]: please reorder imports to codestyle

File: src/main/java/com/icthh/xm/ms/otp/service/LoginPageRefreshableConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+@Component
+public class LoginPageRefreshableConfiguration implements RefreshableConfiguration {
+
+    private final ApplicationProperties appProps;
+    @Getter private String loginHtmlContent;
+
+    @Override
+    public void onRefresh(String updatedKey, String config) {
+        log.info("Setting login page");
+        loginHtmlContent = config;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/otp/service/LoginPageRefreshableConfiguration.java]: how about multi tenancy?

File: src/main/java/com/icthh/xm/ms/otp/service/LoginPageRefreshableConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Override
+    public boolean isListeningConfiguration(String updatedKey) {
+        return Optional.ofNullable(appProps.getLoginPage()).equals(Optional.of(updatedKey));
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/otp/service/LoginPageRefreshableConfiguration.java]: how about multi tenancy support?

File: src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java [hunk 4]
```diff
@@ -75,4 +103,56 @@ public OneTimePasswordResource(OneTimePasswordService oneTimePasswordService) {
...
+    @GetMapping("/userinfo")
+    @Timed
+    public ResponseEntity getUserInfo() {
+        Optional<String> login = authenticationContextHolder.getContext().getAdditionalDetailsValue(MSISDN);
+        if (!login.isPresent()) {
+            log.warn("MSISDN is not set in access token");
+            return ResponseEntity.badRequest().build();
+        }
+        String phoneNumber = login.get();
+        return ResponseEntity.ok(new HashMap<String, String>(){{
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java]: maybe better to user ImmutableMap.of
and static import?
of("phoneNumber", phoneNumber, "id", phoneNumber)

File: src/main/resources/config/application.yml [hunk 5]
```diff
@@ -149,6 +149,7 @@ xm-config:
 # ===================================================================
 
 application:
+    login-page: /config/tenants/XM/otp/one-time-password-login.htm
```
Comment (sergeysenja1992) [inline src/main/resources/config/application.yml]: better to user template, for multi tenancy support

File: src/test/java/com/icthh/xm/ms/otp/repository/UaaRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Qualifier("internalRestTemplate")
+    @MockBean
+    private RestTemplate restTemplate;
+
+
+    @Test
+    public void oauthTokenShouldBeRetrievedFromUaaTest() {
+
+        when(restTemplate.exchange(anyString(), eq(HttpMethod.POST), any(HttpEntity.class), any(ParameterizedTypeReference.class)))
```
Comment (sergeysenja1992) [inline src/test/java/com/icthh/xm/ms/otp/repository/UaaRepositoryTest.java]: not mock restTemplate
better use 
org.springframework.test.web.client.MockRestServiceServer

Comment (sergeysenja1992) [reply src/test/java/com/icthh/xm/ms/otp/repository/UaaRepositoryTest.java]: RestGatewaySupport gateway = new RestGatewaySupport();
        gateway.setRestTemplate(restTemplate);
        mockServer = MockRestServiceServer.createServer(gateway);

File: src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java [hunk 7]
```diff
@@ -1,37 +1,65 @@
...
+    @GetMapping(value = "/login", produces = MediaType.TEXT_HTML_VALUE)
+    public ResponseEntity<String> login() {
 
-    public OneTimePasswordResource(OneTimePasswordService oneTimePasswordService) {
-        this.oneTimePasswordService = oneTimePasswordService;
+        if (loginPageRefreshableConfiguration.getLoginHtmlContent() == null){
+            return ResponseEntity.notFound().build();
+        }
+        log.debug("REST login request");
+        return ResponseEntity.ok(loginPageRefreshableConfiguration.getLoginHtmlContent());
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java]: it's incorrect
you must use method without @ReponseBody (now it inside RestController annotation) and return just template name

File: src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java [hunk 8]
```diff
@@ -75,4 +103,56 @@ public OneTimePasswordResource(OneTimePasswordService oneTimePasswordService) {
...
+    @PreAuthorize("hasPermission({'oneTimePasswordCheckDto': #oneTimePasswordCheckDto} ,'OTP.ONETIMEPASSWORD.CHECK')")
+    @PostMapping("/one-time-password/validate")
+    @Timed
+    public RedirectView checkOneTimePasswordAndRedirectWithCode(@Valid @RequestBody OneTimePasswordCheckDto oneTimePasswordCheckDto,
+                                                                @RequestHeader(value = "redirect-uri") String redirectUri) throws URISyntaxException {
+        redirectUri = decodeUrl(redirectUri);
+        checkOneTimePassword(oneTimePasswordCheckDto);
+        OneTimePasswordDto dto = oneTimePasswordService.findOne(oneTimePasswordCheckDto.getId()).get();
+        String code =  uaaRepository.getOAuth2Token(ImmutableMap.of(MSISDN, dto.getReceiver()));
+        return new RedirectView(new StringBuilder(redirectUri).append("?code=").append(code).toString(), true);
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java]: mmm, how about method setAttributes ?

File: src/main/java/com/icthh/xm/ms/otp/repository/UaaRepository.java (hunk 1 above)
Comment (ibilous) [reply src/main/java/com/icthh/xm/ms/otp/repository/UaaRepository.java]: what code style ?

File: src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java (hunk 7 above)
Comment (ibilous) [reply src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java]: how come i should return template name if i get html string  from config service ?

File: src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java (hunk 8 above)
Comment (ibilous) [reply src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java]: and what about `setAttributes()` ? Yes we can set http status code via setAttributes and use e.g ModelAndView for redirection ? Is it better ?

File: src/main/java/com/icthh/xm/ms/otp/config/Constants.java [hunk 9]
```diff
@@ -12,7 +12,10 @@
     public static final String LOGIN_REGEX = "^[_.@A-Za-z0-9-]*$";
 
     public static final String SYSTEM_ACCOUNT = "system";
+    public static final String X_TENANT = "x-tenant";
+    public static final String ACCESS_TOKEN = "access_token";
     public static final String ANONYMOUS_USER = "anonymoususer";
+    public static final String MSISDN = "msisdn";
```
Comment (ishkurko) [inline src/main/java/com/icthh/xm/ms/otp/config/Constants.java]: Better to use a more universal name, like  `receiver` - in future, it also could be an email, IP, etc.

File: src/main/java/com/icthh/xm/ms/otp/security/HttpComponentsClientHttpRequestFactoryBasicAuth.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import org.apache.http.impl.client.BasicAuthCache;
+import org.apache.http.protocol.BasicHttpContext;
+import org.apache.http.protocol.HttpContext;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
+
+import java.net.URI;
+
+public class HttpComponentsClientHttpRequestFactoryBasicAuth extends HttpComponentsClientHttpRequestFactory {
+
```
Comment (ishkurko) [inline src/main/java/com/icthh/xm/ms/otp/security/HttpComponentsClientHttpRequestFactoryBasicAuth.java]: Plz, do not use double and more empty lines.

File: src/main/java/com/icthh/xm/ms/otp/service/LoginPageRefreshableConfiguration.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.stereotype.Service;
+import org.springframework.util.AntPathMatcher;
+
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Slf4j
+@RequiredArgsConstructor
+@Component
+public class LoginPageRefreshableConfiguration implements RefreshableConfiguration {
```
Comment (ishkurko) [inline src/main/java/com/icthh/xm/ms/otp/service/LoginPageRefreshableConfiguration.java]: Why this class in the service package? Maybe better to move it in the config?

File: src/main/java/com/icthh/xm/ms/otp/repository/UaaRepository.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import static com.icthh.xm.ms.otp.config.Constants.X_TENANT;
+
+@Component
+@RequiredArgsConstructor
+public class UaaRepository {
```
Comment (ishkurko) [inline src/main/java/com/icthh/xm/ms/otp/repository/UaaRepository.java]: `Repository` name and location is the correct thing for this class?

File: src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java [hunk 13]
```diff
@@ -1,37 +1,66 @@
...
+    @Timed
+    @GetMapping(value = "/login", produces = MediaType.TEXT_HTML_VALUE)
+    public ResponseEntity<String> login() {
 
-    public OneTimePasswordResource(OneTimePasswordService oneTimePasswordService) {
-        this.oneTimePasswordService = oneTimePasswordService;
+        if (loginPageRefreshableConfiguration.getLoginContent() == null){
+            return ResponseEntity.notFound().build();
+        }
+        log.debug("REST login request");
```
Comment (ishkurko) [inline src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java]: Why we need this logging? We have request and response logging.

File: src/main/java/com/icthh/xm/ms/otp/config/SecurityConfiguration.java [hunk 14]
```diff
@@ -44,6 +48,9 @@ public void configure(HttpSecurity http) throws Exception {
             .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
         .and()
             .authorizeRequests()
+            .antMatchers("/api/login").permitAll()
+            .antMatchers("/api/userinfo").permitAll()
+            .antMatchers("/api/oauth/token").permitAll()
```
Comment (ishkurko) [inline src/main/java/com/icthh/xm/ms/otp/config/SecurityConfiguration.java]: Why do we permiAll for this endpoint?

File: src/main/java/com/icthh/xm/ms/otp/repository/UaaRepository.java (hunk 1 above)
Comment (sergeysenja1992) [reply src/main/java/com/icthh/xm/ms/otp/repository/UaaRepository.java]: @ishkurko  please send link to code style

File: src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java (hunk 7 above)
Comment (sergeysenja1992) [reply src/main/java/com/icthh/xm/ms/otp/web/rest/OneTimePasswordResource.java]: add implementation of class View that return html from config in render method
and add ViewResolver that return you View by name

File: src/main/java/com/icthh/xm/ms/otp/service/LoginPageRefreshableConfiguration.java (hunk 11 above)
Comment (ibilous) [reply src/main/java/com/icthh/xm/ms/otp/service/LoginPageRefreshableConfiguration.java]: there is already implemented class `OtpSpecService` that in fact does similar - retrieving configuration from config service and map it to POJOs. I think `...Configuration ` is more suitable name for such classes  but i did not rename already existed class and decided to put my new one near  the old one

File: src/main/java/com/icthh/xm/ms/otp/config/SecurityConfiguration.java (hunk 14 above)
Comment (ibilous) [reply src/main/java/com/icthh/xm/ms/otp/config/SecurityConfiguration.java]: i assume we cannot avoid permitting api/login coz user does not have any permission/token before interacting with our system. If i remove next 2 endpoints, add 2 @PreAuthorize annotations and give necessary permissions on web UI i get 401 when uaa communicating with otp. Looks like these requests require some additional token(s). From the other hand am not sure that these endpoints  should be protected: first just convert `code` parameter into `access_token` and second  retrieves user data from jwt token and sends it back to uaa.

**Pull request merged**